src/par_iter/collect/consumer.rs

```diff
@@ -2,50 +2,39 @@ use super::super::len::*;
 use super::super::internal::*;
 use super::super::noop::*;
 use std::ptr;
+use std::slice;
 use std::sync::atomic::{AtomicUsize, Ordering};
 
-pub struct CollectConsumer<'c, ITEM: Send> {
+pub struct CollectConsumer<'c, ITEM: Send + 'c> {
     /// Tracks how many items we successfully wrote. Used to guarantee
     /// safety in the face of panics or buggy parallel iterators.
     writes: &'c AtomicUsize,
 
-    target: *mut ITEM,
-
-    /// Invariant: we can safely write from `target` .. `target+len`
-    len: usize,
+    /// A slice covering the target memory, not yet initialized!
+    target: &'c mut [ITEM],
```

Comment:
It also lets us drop the `isize::MAX` limit, though I doubt that was ever a practical problem, and we get to use whatever the standard library decides is the optimal way to walk through the slice.  My performance testing on `vec_collect` actually looked a little favorable here, but it varied enough from run to run that I didn't want to quote that.